### ISSUE TITLE
TestValidateOnlyStatus flakes with `apiVersion: Invalid value`

### ISSUE THREAD

**Issue by liggitt** [kubernetes/kubernetes#72493]
https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-integration-master/1135464872302088192
```
--- FAIL: TestValidateOnlyStatus (1.61s)
I0102 19:03:43.523081  125200 establishing_controller.go:84] Shutting down EstablishingController
I0102 19:03:43.523219  125200 secure_serving.go:156] Stopped listening on 127.0.0.1:46393
I0102 19:03:43.523225  125200 naming_controller.go:295] Shutting down NamingConditionController
I0102 19:03:43.523093  125200 crd_finalizer.go:254] Shutting down CRDFinalizer
I0102 19:03:43.523081  125200 customresource_discovery_controller.go:214] Shutting down DiscoveryController
testserver.go:141: runtime-config=map[api/all:true]
testserver.go:142: Starting apiextensions-apiserver on port 36085...
testserver.go:160: Waiting for /healthz to be ok...
subresources_test.go:531: unexpected error: WishIHadChosenNoxu.mygroup.example.com "foo" is invalid: apiVersion: Invalid value: "mygroup.example.com/v1beta1": must be mygroup.example.com/v1
```

cc @mbohlool @sttts

**Comment by tallclair** [comment]
https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/72761/pull-kubernetes-integration/40842/

**Comment by liggitt** [comment]
adding to the milestone for investigation

**Comment by liggitt** [comment]
still one of the top 3 integration flakes:
http://storage.googleapis.com/k8s-metrics/flakes-latest.json

https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&test=TestValidateOnlyStatus

**Comment by liggitt** [comment]
the status strategy stomps whatever comes from the incoming object with the objectmeta and non-status fields here:
https://github.com/kubernetes/kubernetes/blob/3cd41a5a0bc4bbdf156bf925316713a1c41d634a/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/status_strategy.go#L35-L54

it then validates the resulting object here:
https://github.com/kubernetes/kubernetes/blob/3cd41a5a0bc4bbdf156bf925316713a1c41d634a/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/status_strategy.go#L56-L59
https://github.com/kubernetes/kubernetes/blob/3cd41a5a0bc4bbdf156bf925316713a1c41d634a/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go#L115-L117
https://github.com/kubernetes/kubernetes/blob/3cd41a5a0bc4bbdf156bf925316713a1c41d634a/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go#L140-L142

if that fails, I *think* that means oldObject is not in the same version as the incoming request, which is significant problem

**Comment by liggitt** [comment]
> if that fails, I think that means oldObject is not in the same version as the incoming request, which is significant problem

The error (`Invalid value: "mygroup.example.com/v1beta1": must be mygroup.example.com/v1`) seems to match that.

**Comment by liggitt** [comment]
Found a consistent reproducer:

1. Create a CRD with two versions and a status subresource (this also gives write permission anonymously so curl works easily):
```
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: foos.example.com
spec:
  group: example.com
  scope: Namespaced
  versions:
  - name: v1
    storage: true
    served: true
  - name: v2
    storage: false
    served: true
  subresources:
    status: {}
  names:
    singular: foo
    plural: foos
    kind: Foo
    listKind: FooList
---
apiVersion: example.com/v1
kind: Foo
metadata:
  name: foo
spec:
  specValue: true
status:
  specValue: true
---
kind: Role
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: foo-role
rules:
- apiGroups: ["example.com"]
  resources: ["foos", "foos/status"]
  verbs: ["*"]
---
kind: RoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: foo-role-binding
roleRef:
  kind: Role
  apiGroup: rbac.authorization.k8s.io
  name: foo-role
subjects:
- kind: Group
  name: system:authenticated
- kind: Group
  name: system:unauthenticated
```

2. Set up two loops, each patching the status subresource as fast as they can, each on a different version

tab 1:
```sh
for i in $( seq 1 1000 ); do
    echo $i
    v1result=$(curl -s -X PATCH https://localhost:6443/apis/example.com/v1/namespaces/default/foos/foo/status -H "Content-Type: application/merge-patch+json" --data '{"status":{"v1":"'$i'"}}' -k)
    echo "${v1result}"
    if [[ "${v1result}" == *"422"* ]]; then
        break
    fi
done
```

tab 2:
```
for i in $( seq 1 1000 ); do
    echo $i
    v2result=$(curl -s -X PATCH https://localhost:6443/apis/example.com/v2/namespaces/default/foos/foo/status -H "Content-Type: application/merge-patch+json" --data '{"status":{"v2":"'$i'"}}' -k)
    echo "${v2result}"
    if [[ "${v2result}" == *"422"* ]]; then
        break
    fi
done
```

very quickly, this is triggered from one of the tabs, complaining about the opposite version:
```
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "Foo.example.com \"foo\" is invalid: apiVersion: Invalid value: \"example.com/v1\": must be example.com/v2",
  "reason": "Invalid",
  "details": {
    "name": "foo",
    "group": "example.com",
    "kind": "Foo",
    "causes": [
      {
        "reason": "FieldValueInvalid",
        "message": "Invalid value: \"example.com/v1\": must be example.com/v2",
        "field": "apiVersion"
      }
    ]
  },
  "code": 422
}
```

**Comment by liggitt** [comment]
raising priority since this is a real bug, not just a test flake

**Comment by roycaihw** [comment]
looking now

**Comment by liggitt** [comment]
disabling the watch cache (starting the apiserver with --watch-cache=false) makes it fail 100% of the time on the non-storage version

**Comment by liggitt** [comment]
that means that the non-cache storage is not converting the old object into the correct in-memory version for the handler

**Comment by liggitt** [comment]
figured it out... statusREST is not setting the expected kind in the New() func, so conversion doesn't know a different in-memory version is expected.

**Comment by liggitt** [comment]
it looks like StatusREST#Get also needs the same fix as the main REST handler for shallow copying object meta

**Comment by liggitt** [comment]
even with those fixed, the oldObject being sent to Update methods is not being converted. updated the conversion integration test to trigger this and to run a set of tests with the watch cache disable to ensure we function correctly. still digging into why the conversion isn't doing what we expect

**Comment by liggitt** [comment]
found the root cause... the etcd store code that creates the new object to decode into was not propagating the desired hub group version kind:

https://github.com/kubernetes/kubernetes/blob/714fcd910fb7abf8647a6f67bfb15eeed10854a1/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L687-L690

That meant that when reading from etcd, the "old object" was always kept in whatever version it was read from storage in, rather than being converted to the in-memory hub version.

This was masked by several things:
* In almost all updates, the "current object" comes from the last-observed update in the REST version held by the watch cache, and is not read from etcd. That is why this test only rarely flaked, but failed 100% of the time with the watch cache disabled
* For PUT and POST requests to the primary resource (not a subresource), the apiVersion of the old object never hits validation... only in patch requests can the fields from the oldObject bleed through into validation
* For all requests to the status subresource of a custom resource, all fields from the old object except "status" are preserved. That's why this was the test that flaked when we happened to get an unconverted object from etcd rather than the watch cache.

Fixed the issue in https://github.com/kubernetes/kubernetes/pull/78713 and added tests to exercise the code path to ensure the old object has the correct GVK with and without the watch cache enabled
